File: 95.unique-binary-search-trees-ii.cpp
/*
 * @lc app=leetcode id=95 lang=cpp
 *
 * [95] Unique Binary Search Trees II
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    std::vector<TreeNode*> constructTrees(int start, int end) {
        std::vector<TreeNode*> list;

        if (start > end) {
            list.push_back(nullptr);
            return list;
        }

        for (int i = start; i <= end; i++) {
            std::vector<TreeNode*> leftSubtree = constructTrees(start, i - 1);
            std::vector<TreeNode*> rightSubtree = constructTrees(i + 1, end);

            for (TreeNode* left : leftSubtree) {
                for (TreeNode* right : rightSubtree) {
                    TreeNode* node = new TreeNode(i);
                    node->left = left;
                    node->right = right;
                    list.push_back(node);
                }
            }
        }
        return list;
    }

    std::vector<TreeNode*> generateTrees(int n) {
        return constructTrees(1, n);
    }
};
// @lc code=end

File: 95.unique-binary-search-trees-ii.py
#
# @lc app=leetcode id=95 lang=python3
#
# [95] Unique Binary Search Trees II
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from functools import lru_cache
from typing import List, Optional


class Solution:
    def constructTrees(self, start, end):
        list = []
        if (start > end) :
            list.append(None)
            return list
        for i in range(start, end + 1):
            leftSubtree = self.constructTrees(start, i - 1)
            rightSubtree = self.constructTrees(i + 1, end)

            for j in range(len(leftSubtree)) :
                left = leftSubtree[j]
                for k in range(len(rightSubtree)):
                    right = rightSubtree[k]
                    node = TreeNode(i)   
                    node.left = left    
                    node.right = right   
                    list.append(node)    
        return list

    def generateTrees(self, n: int) -> List[Optional[TreeNode]]:
        return self.constructTrees(1,n)
        
# @lc code=end

